src/compiler/compiler.rs

```diff
@@ -556,41 +560,44 @@ gcc
             //TODO: do something smarter here.
             if line == "gcc" {
                 debug!("Found GCC");
-                return future::ok(Some(CompilerKind::Gcc)).boxed()
+                return future::ok(Some(Box::new(CCompiler(GCC)) as Box<Compiler<T>>)).boxed()
             } else if line == "clang" {
                 debug!("Found clang");
-                return future::ok(Some(CompilerKind::Clang)).boxed()
+                return future::ok(Some(Box::new(CCompiler(Clang)) as Box<Compiler<T>>)).boxed()
             } else if line == "msvc" {
                 debug!("Found MSVC");
                 let prefix = msvc::detect_showincludes_prefix(&creator,
                                                               executable.as_ref(),
                                                               &pool);
                 return Box::new(prefix.map(|prefix| {
                     trace!("showIncludes prefix: '{}'", prefix);
-                    Some(CompilerKind::Msvc {
+                    Some(Box::new(CCompiler(MSVC {
                         includes_prefix: prefix,
-                    })
+                    })) as Box<Compiler<T>>)
                 }))
             }
         }
         future::ok(None).boxed()
     }))
 }
 
-/// If `executable` is a known compiler, return `Some(Compiler)` containing information about it.
+/// If `executable` is a known compiler, return a `CompilerInfo` containing information about it.
 pub fn get_compiler_info<T>(creator: &T, executable: &str, pool: &CpuPool)
-                            -> SFuture<Compiler>
+                            -> SFuture<CompilerInfo<T>>
     where T: CommandCreatorSync
 {
     let executable = executable.to_string();
     let pool = pool.clone();
-    Box::new(detect_compiler_kind(creator, &executable, &pool).and_then(move |kind| {
-        match kind {
-            Some(kind) => {
-                pool.spawn_fn(move || Compiler::new(&executable, kind))
-                    .chain_err(|| "failed to learn compiler metadata")
+    let detect = detect_compiler(creator, &executable, &pool);
+    Box::new(detect.and_then(move |compiler| {
+        match compiler {
+            Some(compiler) => {
+                Box::new(sha1_digest(executable.clone(), &pool)
```

Comment:
This is actually somewhat interesting in the rustc case. I'd imagine that quite commonly this is actually the rustup version of `rustc.exe` rather than the rustc version of `rustc.exe`. I wonder if we could automatically detect rustup and go fetch the actual compiler instead?

I believe that would work best by looking for `rustup.exe` in the directory next to `rustc.exe` (if found). If that's there we can run `rustup which rustc` which should print out the full path to the actual compiler.

Also as a side note here, I wonder if we should handle the dynamic libraries that rustc itself links to? I can't really imagine a use case though where the dylibs are changed and the rustc executable *wouldn't* change, though, so it's perhaps not an issue.